lib/internal/observable.js

```diff
@@ -0,0 +1,898 @@
+ * @param {Subscriber} subscriber
+ */
+
+/**
+ * @callback SubscriptionObserverCallback
+ * @param {any} value
+ * @returns {void}
+ */
+
+/**
+ * @callback VoidFunction
+ * @returns {void}
+ */
+
+/**
+ * @typedef {Object} SubscriptionObserver
+ * @property {SubscriptionObserverCallback} [next]
+ * @property {SubscriptionObserverCallback} [error]
+ * @property {VoidFunction} [complete]
+ */
+
+/**
+ * @callback Mapper
+ * @param {any} element
+ * @param {number} index
+ * @returns {any}
+ */
+
+/**
+ * @callback Predicate
+ * @param {any} element
+ * @param {number} index
+ * @returns {boolean}
+ */
+
+/**
+ * @callback Reducer
+ * @param {any} accumulator
+ * @param {any} element
+ * @returns {any}
+ */
+
+class Subscriber {
+  constructor() {
+    throw new ERR_ILLEGAL_INVOCATION('Subscriber');
+  }
+
+  next(value) { this[kState]?.next(value); }
+
+  complete() { this[kState]?.complete(); }
+
+  error(error) {
+    const err = this[kState]?.error;
+    this[kClose]();
+    if (typeof err === 'function') {
+      err(error);
+    } else {
+      throw error;
+      // The spec expects us to use reportError();
+    }
+  }
+
+  addTeardown(teardown) {
+    if (this[kState] === undefined) return;
+    this[kState].teardowns.push(teardown);
+  }
+
+  get active() { return this[kState] === undefined; }
+
+  get signal() { return this[kSignal]; }
+
+  [kClose]() { this[kState] = undefined; }
+
+  [kTeardown]() {
+    if (this[kState] === undefined) return;
+    const teardowns = this[kState].teardowns;
+    this[kState] = undefined;
+    teardowns.reverse();
+    for (const teardown of teardowns) {
+      teardown();
+    }
+  }
+}
+
+/**
+ * @param {SubscribOptions} options
+ * @returns {Subscriber}
+ */
+function createSubscriber(internalObserver, options) {
+  validateObject(options, 'options');
+  validateObject(internalObserver, 'internalObserver');
+  const { signal } = options;
+  const { next, error, complete } = internalObserver;
+  if (signal !== undefined) {
+    validateAbortSignal(signal, 'options.signal');
+  }
+  return ReflectConstruct(function() {
+    this[kState] = {
+      completeOrErrorController: new AbortController(),
+      teardowns: [],
+      next,
+      error,
+      complete,
+    };
+    const signals = [
+      this[kState].completeOrErrorController.signal
+    ];
+    if (signal !== undefined) signals.push(signal);
+    this[kSignal] = AbortSignal.any(signals);
+  }, [], Subscriber);
+}
+
+class Observable {
+  #subscriberCallback = undefined;
+
+  static from(value) {
+    if (value?.[kObserverMarker] === kObserverMarker) return value;
+
+    if (value instanceof Promise) {
+      return new Observable((subscriber) => {
+        value.then(
+          (value) => {
+            subscriber.next(value);
+            subscriber.complete();
+          },
+          (error) => subscriber.error(error)
+        );
+      });
+    }
+
+    if (typeof value?.[SymbolAsyncIterator] === 'function' ||
```

Comment:
@benlesh I'm surprised Observable.from consumes async iterables but observables aren't async iterable in this iteration.